nixos/virtualisation: add option for explicitly named network interfaces

andrew-hoff opened the pull request:
Adds a new option to the virtualisation modules that enables specifying explicitly named network interfaces in QEMU VMs. The existing `virtualisation.vlans` is still supported for cases where the name of the network interface is irrelevant.

This feature is useful for creating VMs with network configurations that exactly match the network configuration of real-world devices for more accurate SITL testing.

###### Description of changes

- Added `virtualisation.interfaces` option to `qemu-vm.nix`.
- Updated the default value of `virtualisation.vlans` to empty if `virtualisation.interfaces` is not empty. This avoids a scenario where a user specifies `virtualisation.interfaces`, but inadvertently still has a network interface from the default `virtualisation.vlans`.
- Updated `build-vms.nix` to create the new network interfaces using QEMU options.
- Updated `build-vms.nix` to rename network interfaces using udev rules.
- Updated `testing-python.nix` to use the union of the VLANs specified in `virtualisation.vlans` and `virtualisation.interfaces` when collecting the list of VLANs.
- Updated networking tests in `nixos/tests/networking.nix` to use the new option where applicable. This cut down on unnecessary overrides to remove IP addresses from network interfaces.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

andrew-hoff commented on nixos/lib/build-vms.nix:
```diff
@@ -58,24 +58,42 @@ rec {
 
+              qemuOptions = flatten (forEach interfacesNumbered ({ fst, snd }:
+                qemu-common.qemuNICFlags snd fst.vlan m.snd));
+              udevRules = forEach interfacesNumbered ({ fst, snd }:
+                "SUBSYSTEM==\"net\",ACTION==\"add\",ATTR{address}==\"${qemu-common.qemuNicMac fst.vlan m.snd}\",NAME=\"${fst.name}\"");
```
I chose to use udev rules rather than `systemd.network.links` because in the latter case, systemd ignores all but one file that applies to a given interface. I did not want to preclude users from using that module elsewhere.

andrew-hoff commented:
cc @roberth since I see you as the last committer to build-vms.nix

roberth commented:
Oh, I'm actually replacing `build-vms.nix` by modules in

 - https://github.com/NixOS/nixpkgs/pull/176557

The changes in `build-vms.nix` can be applied to the `testing/network.nix` file instead.

I'm arranging to run aforementioned PR on Hydra. I think this PR can be the next one to run there.

roberth commented on nixos/modules/virtualisation/qemu-vm.nix:
```diff
@@ -511,7 +511,7 @@ in
       mkOption {
         type = types.listOf types.ints.unsigned;
-        default = [ 1 ];
+        default = if config.virtualisation.interfaces == {} then [ 1 ] else [ ];
```
```suggestion
        default = if config.virtualisation.interfaces == {} then [ 1 ] else [ ];
        defaultText = lib.literalExpression ''if config.virtualisation.interfaces == {} then [ 1 ] else [ ]'';
```

andrew-hoff commented:
> 

Got it. Should I wait until that gets merged and rebase? Or should I just cut a branch off of yours?

roberth commented:
tl;dr: This PR is **not blocked** and can be reviewed by any maintainer who is familiar with NixOS networking.

> > Oh, I'm actually replacing `build-vms.nix` by modules in
> > ```
> > * [nixosTest: modularity and matrix #176557](https://github.com/NixOS/nixpkgs/pull/176557)
> > ```
> >   
> > The changes in `build-vms.nix` can be applied to the `testing/network.nix` file instead.
> > I'm arranging to run aforementioned PR on Hydra. I think this PR can be the next one to run there.
> 
> Got it. Should I wait until that gets merged and rebase? Or should I just cut a branch off of yours?

Ok, while the big refactor is in a good shape, we won't be able to get it properly reviewed and merged in the coming 3 weeks.
The listed maintainer is occupied during this time, and as for myself, I already went far over my time budget with my own contributions as well, so this PR isn't going to receive a review from the two of us until around that time at least.

That all said, this PR isn't anything that can't be ported to my branch, and if another maintainer can review and merge this in the meanwhile, that is fine by me. I don't see any issues with how the code is structured.

roberth commented:
Perhaps @fpletz could review the changes, as it touches the `networking.nix` test?

We can also run all tests on hydra after the current https://hydra.nixos.org/jobset/nixos/python-test-refactoring completes.

andrew-hoff commented:
I just opened up a new PR against your branch [here](https://github.com/hercules-ci/nixpkgs/pull/1). I'll add @roberth and @fpletz as reviewers, though I completely understand if you can't get to it right away.

roberth commented:
Hi, the module changes have been merged, so you can rebase the PR.

I was looking for a qemu networking expert just now. What do you think of https://github.com/NixOS/nixpkgs/pull/200225?

RaitoBezarius commented:
I am interested into helping getting this merged, can you fix the conflicts @andrew-hoff and move the release notes to 23.05 ?

andrew-hoff committed:
nixos/virtualisation: add option for explicitly named network interfaces

Adds a new option to the virtualisation modules that enables specifying
explicitly named network interfaces in QEMU VMs. The existing
`virtualisation.vlans` is still supported for cases where the name of
the network interface is irrelevant.

andrew-hoff commented:
@RaitoBezarius I updated the PR and moved release notes to 23.05. Sorry for the slow response!

RaitoBezarius commented:
@andrew-hoff Thank you for maintaining this conflict-less, it seems you introduced a merge commit recently, can you get rid of it?
I think if we don't have a review, I would merge this and we can always revert it in case of problems.

andrew-hoff commented:
> @andrew-hoff Thank you for maintaining this conflict-less, it seems you introduced a merge commit recently, can you get rid of it? I think if we don't have a review, I would merge this and we can always revert it in case of problems.

Ah, sorry about that. Just cleaned it up.

andrew-hoff commented:
Thank you @RaitoBezarius !

RaitoBezarius commented:
> Thank you @RaitoBezarius !

Thank YOU for your incredible patience. :)

vcunat commented:
https://github.com/NixOS/nixpkgs/pull/213118
Can you perhaps look again if other things could go wrong when resurrecting the PR after months?

RaitoBezarius commented:
> #213118 Can you perhaps look again if other things could go wrong when resurrecting the PR after months?

Apologies for that breakage, I will run more tests to check.

Artturin commented on nixos/tests/networking.nix:
```diff
@@ -947,30 +933,26 @@ let
         print(machine.succeed("ip link show dev custom_name"))
       '';
-    };
+    } else {
+      name = "RenameInterface";
+      testScript = "";
```
this is unfinished

```
> nixosTests.networking.scripted.rename
error: The option `nodes' is used but not defined.
```

raphaelr replied:
Fixed by https://github.com/NixOS/nixpkgs/pull/213371

vcunat commented:
Well, while _evaluation_ got fixed, some of the tests fail or time out.  (And they work if I revert this PR and the fixups.)  And so the `nixos-unstable` channel is still blocked.  https://hydra.nixos.org/build/207314144#tabs-constituents

RaitoBezarius commented:
> Well, while _evaluation_ got fixed, some of the tests fail or time out. (And they work if I revert this PR and the fixups.) And so the `nixos-unstable` channel is still blocked. https://hydra.nixos.org/build/207314144#tabs-constituents

I'm fine with reverting this PR as it requires clearly more work at the moment I guess.

vcunat commented:
Opened as https://github.com/NixOS/nixpkgs/pull/213519
(if noone has a fix)
